### ISSUE TITLE
Practical 2B: XOR Function using Multi-Layer Perceptron

### ISSUE THREAD

**Issue by karinkato2103** [karinkato2103/soft#4]
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

class XOR_MLP:
    def __init__(self):
        self.lr = 0.5
        self.epochs = 20000
        rng = np.random.RandomState(1)

        self.w1 = rng.normal(scale=0.5, size=(2, 2))
        self.b1 = np.zeros(2)

        self.w2 = rng.normal(scale=0.5, size=(2, 1))
        self.b2 = np.zeros(1)

    def train(self, X, y):
        for _ in range(self.epochs):
            z1 = X.dot(self.w1) + self.b1
            a1 = sigmoid(z1)

            z2 = a1.dot(self.w2) + self.b2
            a2 = sigmoid(z2)

            error = a2 - y.reshape(-1, 1)

            d2 = error * a2 * (1 - a2)
            dw2 = a1.T.dot(d2)
            db2 = d2.sum(axis=0)

            d1 = d2.dot(self.w2.T) * a1 * (1 - a1)
            dw1 = X.T.dot(d1)
            db1 = d1.sum(axis=0)

            self.w2 -= self.lr * dw2
            self.b2 -= self.lr * db2

            self.w1 -= self.lr * dw1
            self.b1 -= self.lr * db1

    def predict(self, X):
        a1 = sigmoid(X.dot(self.w1) + self.b1)
        a2 = sigmoid(a1.dot(self.w2) + self.b2)
        return (a2 >= 0.5).astype(int)

if __name__ == "__main__":
    X = np.array([[0,0],[0,1],[1,0],[1,1]])
    y = np.array([0,1,1,0])

    model = XOR_MLP()
    model.train(X, y)

    print(model.predict(X).reshape(-1))
